4.09: # stat: can't read file system information for '%m': No such file or directory

I have the Alpine-based Dockerfile, with these lines:
```Dockerfile
FROM alpine:edge AS builder
RUN apk add --no-cache make m4 which patch sudo wget git pkgconfig \
	gcc g++ musl-dev linux-headers libffi libffi-dev pcre perl python3 \
	gmp-dev libressl-dev pcre-dev zlib-dev xz-dev \
	ocaml ocaml-compiler-libs ocaml-ocamldoc opam ocaml-findlib \
	unzip gmp xz

RUN chmod +s `which bwrap`

RUN addgroup -g 1000 -S user && adduser -u 1000 -D -S user -G user
RUN sed -i.bkp -e \
	's/%wheel\s\+ALL=(ALL\(:ALL\)\?)\s\+ALL/%wheel ALL=NOPASSWD:ALL/g' \
	/etc/sudoers
COPY ocamlinit /home/user/.ocamlinit
USER user
WORKDIR /home/user

ENV OCAML_TOPLEVEL_PATH /home/user/.opam/default/lib/toplevel
RUN opam init --auto-setup --disable-sandboxing --yes && \
	# FIXME: Ugly hack
	ln -s /home/user/.opam/default/lib/toplevel /home/user/.opam/default/lib/topfind && \
	eval `opam env` && opam install --yes ocamlfind && \
	opam update --yes && export CHECK_IF_PREINSTALLED=false; \
	opam install --yes opam-devel

RUN eval `opam env` && \
	opam switch create 4.09.0 && \
	opam switch 4.09.0
```
And during the build it returns the following error
```
[ocaml-base-compiler.4.09.0] downloaded from cache at https://opam.ocaml.org/cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
-> installed base-bigarray.base
-> installed base-threads.base
-> installed base-unix.base
[ERROR] The compilation of ocaml-base-compiler failed at "/usr/bin/make -j7 world".

#=== ERROR while compiling ocaml-base-compiler.4.09.0 =========================#
# context              2.0.3 | linux/x86_64 |  | https://opam.ocaml.org#f335c4dc
# path                 ~/.opam/4.09.0/.opam-switch/build/ocaml-base-compiler.4.09.0
# command              /usr/bin/make -j7 world
# exit-code            2
# env-file             ~/.opam/log/ocaml-base-compiler-7-a00279.env
# output-file          ~/.opam/log/ocaml-base-compiler-7-a00279.out
### output ###
# [...]
# ./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I dr[...]
# ./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I dr[...]
# ./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I dr[...]
# stat: can't read file system information for '%m': No such file or directory
# stat: can't read file system information for '%m': No such file or directory
# tools/check-parser-uptodate-or-warn.sh: line 48: arithmetic syntax error
# make[2]: *** [Makefile:1079: parsing/parser.ml] Error 2
# make[2]: *** Waiting for unfinished jobs....
# make[2]: Leaving directory '/home/user/.opam/4.09.0/.opam-switch/build/ocaml-base-compiler.4.09.0'
# make[1]: *** [Makefile:355: coreall] Error 2
# make[1]: Leaving directory '/home/user/.opam/4.09.0/.opam-switch/build/ocaml-base-compiler.4.09.0'
# make: *** [Makefile:463: world] Error 2

<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+- The following actions failed
| - build ocaml-base-compiler 4.09.0
+- 
+- The following changes have been performed (the rest was aborted)
| - install base-bigarray base
| - install base-threads  base
| - install base-unix     base
+- 

<><> ocaml-base-compiler.4.09.0 troubleshooting <><><><><><><><><><><><><><><><>
=> A failure in the middle of the build may be caused by build parallelism
      (enabled by default).
      Please file a bug report at https://github.com/ocaml/ocaml/issues
# Run eval $(opam env) to update the current shell environment
Switch initialisation failed: clean up? ('n' will leave the switch partially installed) [Y/n] n
The command '/bin/sh -c eval `opam env` &&      opam switch create 4.09.0 &&    opam switch 4.09.0' returned a non-zero code: 31
```

--- a/Changes
+++ b/Changes
@@ -257,6 +257,9 @@ Working version
   (Vincent Laviron, review by Pierre Chambart and Leo White,
    report by Aleksandr Kuzmenko)
 
+- #8965, #8979: Alpine build failure caused by check-parser-uptodate-or-warn.sh
+  (Gabriel Scherer and David Allsopp, report by Anton Kochkov)
+
 OCaml 4.09 maintenance branch:
 ------------------------------
 

--- a/Makefile
+++ b/Makefile
@@ -1094,9 +1094,9 @@ parsing/camlinternalMenhirLib.mli: boot/menhir/menhirLib.mli
 parsing/parser.ml: boot/menhir/parser.ml parsing/parser.mly \
   tools/check-parser-uptodate-or-warn.sh
 	@-tools/check-parser-uptodate-or-warn.sh
-	cat $< | sed "s/MenhirLib/CamlinternalMenhirLib/g" > $@
+	sed "s/MenhirLib/CamlinternalMenhirLib/g" $< > $@
 parsing/parser.mli: boot/menhir/parser.mli
-	cat $< | sed "s/MenhirLib/CamlinternalMenhirLib/g" > $@
+	sed "s/MenhirLib/CamlinternalMenhirLib/g" $< > $@
 
 beforedepend:: parsing/camlinternalMenhirLib.ml \
   parsing/camlinternalMenhirLib.mli \

--- a/Makefile
+++ b/Makefile
@@ -1093,7 +1093,7 @@ parsing/camlinternalMenhirLib.mli: boot/menhir/menhirLib.mli
 
 parsing/parser.ml: boot/menhir/parser.ml parsing/parser.mly \
   tools/check-parser-uptodate-or-warn.sh
-	@tools/check-parser-uptodate-or-warn.sh || true
+	@-tools/check-parser-uptodate-or-warn.sh
 	cat $< | sed "s/MenhirLib/CamlinternalMenhirLib/g" > $@
 parsing/parser.mli: boot/menhir/parser.mli
 	cat $< | sed "s/MenhirLib/CamlinternalMenhirLib/g" > $@

--- a/Makefile
+++ b/Makefile
@@ -1093,7 +1093,7 @@ parsing/camlinternalMenhirLib.mli: boot/menhir/menhirLib.mli
 
 parsing/parser.ml: boot/menhir/parser.ml parsing/parser.mly \
   tools/check-parser-uptodate-or-warn.sh
-	@tools/check-parser-uptodate-or-warn.sh
+	@tools/check-parser-uptodate-or-warn.sh || true
 	cat $< | sed "s/MenhirLib/CamlinternalMenhirLib/g" > $@
 parsing/parser.mli: boot/menhir/parser.mli
 	cat $< | sed "s/MenhirLib/CamlinternalMenhirLib/g" > $@

--- a/tools/check-parser-uptodate-or-warn.sh
+++ b/tools/check-parser-uptodate-or-warn.sh
@@ -15,6 +15,12 @@
 #*                                                                        *
 #**************************************************************************
 
+# stop early if we are not on a development version
+if test -z "$(cat VERSION | grep '+dev')"
+then
+    exit 0
+fi
+
 # We try to warn if the user edits parsing/parser.mly but forgets to
 # rebuild the generated parser. Our heuristic is to use the file
 # modification timestamp, but just testing

--- a/tools/check-parser-uptodate-or-warn.sh
+++ b/tools/check-parser-uptodate-or-warn.sh
@@ -33,7 +33,7 @@ fi
 # using either
 #  GNU coreutils' stat --format, or
 #  busybox's stat -c, or
-#  BSD/macos stat -f.
+#  BSD/macOS stat -f.
 # Default to 0 if 'stat' is not available.
 
 stat . 2>/dev/null 1>/dev/null

--- a/tools/check-parser-uptodate-or-warn.sh
+++ b/tools/check-parser-uptodate-or-warn.sh
@@ -53,7 +53,12 @@ mtime() {
 # The check itself
 SOURCE_MTIME=$(mtime parsing/parser.mly)
 GENERATED_MTIME=$(mtime boot/menhir/parser.ml)
-if test -n "$SOURCE_MTIME" -a -n "$GENERATED_MTIME" -a "$SOURCE_MTIME" -gt $(( GENERATED_MTIME + 10 ))
+if test -z "$SOURCE_MTIME" -o -z "$GENERATED_MTIME"
+then
+  echo
+  tput setaf 3; tput bold; printf "Warning: "; tput sgr0
+  echo "Failed to check if boot/menhir/parser.ml is up-to-date."
+elif test "$SOURCE_MTIME" -gt $(( GENERATED_MTIME + 10 ))
 then
   echo
   tput setaf 3; tput bold; printf "Warning: "; tput sgr0

--- a/tools/check-parser-uptodate-or-warn.sh
+++ b/tools/check-parser-uptodate-or-warn.sh
@@ -16,10 +16,7 @@
 #**************************************************************************
 
 # stop early if we are not on a development version
-if test -z "$(cat VERSION | grep '+dev')"
-then
-    exit 0
-fi
+grep -Fq '+dev' VERSION || exit 0
 
 # We try to warn if the user edits parsing/parser.mly but forgets to
 # rebuild the generated parser. Our heuristic is to use the file
@@ -39,9 +36,9 @@ fi
 stat . 2>/dev/null 1>/dev/null
 if test $? != 0
 then MTIME=""
-elif test -n "$(stat --version 2>/dev/null | grep coreutils)"
+elif stat --version 2>/dev/null | grep -Fq 'coreutils'
 then MTIME="stat --format %Y"
-elif test -n "$(stat 2>&1 | grep busybox)"
+elif stat 2>&1 | grep -Fq 'busybox'
 then MTIME="stat -c %Y"
 else MTIME="stat -f %m" # BSD stat?
 fi

--- a/tools/check-parser-uptodate-or-warn.sh
+++ b/tools/check-parser-uptodate-or-warn.sh
@@ -50,7 +50,7 @@ mtime() {
 # The check itself
 SOURCE_MTIME=$(mtime parsing/parser.mly)
 GENERATED_MTIME=$(mtime boot/menhir/parser.ml)
-if test "$SOURCE_MTIME" -gt $(( GENERATED_MTIME + 10 ))
+if test -n "$SOURCE_MTIME" -a -n "$GENERATED_MTIME" -a "$SOURCE_MTIME" -gt $(( GENERATED_MTIME + 10 ))
 then
   echo
   tput setaf 3; tput bold; printf "Warning: "; tput sgr0

--- a/tools/check-parser-uptodate-or-warn.sh
+++ b/tools/check-parser-uptodate-or-warn.sh
@@ -24,7 +24,10 @@
 # seconds after boot/menhir/parser.ml.
 
 # mtime(): access a file's last modification time as a timestamp,
-# using either GNU coreutils' stat --format, or BSD/macos stat -f.
+# using either
+#  GNU coreutils' stat --format, or
+#  busybox's stat -c, or
+#  BSD/macos stat -f.
 # Default to 0 if 'stat' is not available.
 
 stat . 2>/dev/null 1>/dev/null
@@ -32,7 +35,9 @@ if test $? != 0
 then MTIME=""
 elif test -n "$(stat --version 2>/dev/null | grep coreutils)"
 then MTIME="stat --format %Y"
-else MTIME="stat -f %m"
+elif test -n "$(stat 2>&1 | grep busybox)"
+then MTIME="stat -c %Y"
+else MTIME="stat -f %m" # BSD stat?
 fi
 
 mtime() {

